File: Server/models/Category.js
import mongoose from 'mongoose';

const categorySchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  name: {
    type: String,
    required: true,
  },
  type: {
    type: String,
    enum: ['Income', 'Expense'],
    required: true,
  },
  icon: {
    type: String, // lucide icon name or emoji
    default: 'Tag',
  },
  color: {
    type: String, // hex code
    default: '#3b82f6',
  },
  parent: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Category',
    default: null,
  },
}, {
  timestamps: true,
});

const Category = mongoose.model('Category', categorySchema);
export default Category;

File: Android/app/src/main/java/com/finora/app/data/model/Category.kt
package com.finora.app.data.model

import com.google.gson.annotations.SerializedName

data class Category(
    @SerializedName("_id") val id: String = "",
    val user: String? = null,
    val name: String = "",
    val type: String = "Expense", // Enum: Income, Expense, Transfer
    val parent: Any? = null, // Can be String ID or populated Category object or null
    val icon: String = "Folder",
    val color: String = "#6b7280",
    val isSystemDefault: Boolean = false,
    val isEssential: Boolean = false,
    val subcategories: List<Category> = emptyList(),
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    val parentId: String?
        get() = when (parent) {
            is String -> parent
            is Map<*, *> -> parent["_id"] as? String
            else -> null
        }
}
